doc-dash-holistic-view (#2)
mmiele/DASH: documentation/general/design/dash-holistic-view.md

    - [Low level architecture](#low-level-architecture)
  - [DASH smart switch](#dash-smart-switch)
    - [High level architecture](#high-level-architecture-1)
    - [Low level architecture](#low-level-architecture-1)
- [Example of network representation](#example-of-network-representation)
  - [Deployment](#deployment)
- [API](#api)
  - [Switch Abstraction Interface (SAI) API](#switch-abstraction-interface-sai-api)
  - [DASH pipeline API](#dash-pipeline-api)
    - [Behavioral model](#behavioral-model)
  - [Functional testing](#functional-testing)
  - [Conformance and performance testing](#conformance-and-performance-testing)
- [Appendix](#appendix)
  - [A day in the life of a DASH packet](#a-day-in-the-life-of-a-dash-packet)
  - [A day in the life of a DASH SDN controller](#a-day-in-the-life-of-a-dash-sdn-controller)
  - [A day in the life of a DASH container](#a-day-in-the-life-of-a-dash-container)
- [References](#references)

## Introduction

This document introduces the **Disaggregated API for SONiC Hosts** (**DASH** for
short) and provides an overview of the project, the supporting elements and its
architecture.

![dash-words-cloud](./images/dash-words-cloud.png)

> DASH is an open source project whose goal is to deliver a **behavioral model**
> and related **test cases**. Technology providers must conform to the model and
> test cases when designing and building network devices. These devices are meant
> to provide **high-speed network connectivity** to critical cloud enterprise
> applications. 

Reviewer: ```suggestion
DASH is an open source project with the goal of delivering a **behavioral model**
and related **test cases** to confirm conformance and performance requirements. Technology providers will conform to the model and
test cases when designing and building programmable hardware technology. These devices are meant
to **deliver enterprise network performance to critical cloud applications**. 

DASH extends SONiC APIs and a related comprehensive set of object models to
describe networking services for the cloud. The project enlists cloud and
enterprise technology providers to collaborate and further extend DASH to meet
their specific needs.
```

DASH extends SONiC APIs and a related comprehensive set of object models that
describe networking services for the cloud. The project enlists cloud and
enterprise technology providers to collaborate and further extend DASH to meet
their specific needs.

### Why DASH? 

> The current technologies in the market by necessity provide general purpose APIs
> to perform general purpose actions (ACLs, forwarding tables, metering, etc.).
> The reason is that it is impossiible to target each specific technology
> provider's implemantation. Obviusely, a general purpose approach can just do so
> much and works against high performance connectivity. 
> 
> DASH takes a different approach. It does not make any assumption or enforce any
> implementation details, such as programming language. It delegates to the
> technology providers such decisions so they can best employ their know-how and
> competitive edge to obtain the best possible results, as long as the agreed upon
> DASH behaviroal model and test cases are satisfied. 

Reviewer: ```suggestion
Because it has been difficult/impossible to target each specific technology provider's implementation(s), the technologies in the market currently provide general purpose APIs to perform general purpose actions (such as ACLs, forwarding tables, metering, etc...).  Again, these are not targeted to specific cloud/enterprise usage.  Obviously, a general purpose approach delivers only so much, and may work against high performance connectivity. 

DASH takes a different approach. DASH is publicizing Behavioral Models to optimize exactly 'what' clouds/enterprises are doing, to optimize against their services. This is the first time 'what we do' has been exposed.  It does not make assumptions or enforce implementation details, such as a programming language. It delegates these decisions to technology providers in order to best employ their expertise and competitive edge to obtain the best possible results (*given that the agreed upon
DASH behavioral model and test cases are satisfied*).
```

### Objectives

The DASH overall objective is to **optimize network SMART Programmable
Technologies performance**, and **leverage commodity hardware technology** to
achieve **10x or even 100x stateful connection performance**. In particular

- With the help of network hardware technology providers, create an open forum
  that capitalizes on the use of **programmable networking devices** including
  - **SmartNICs**. A smart network interface card (SmartNIC), is a programmable

>     accelerator that makes data center networking, security and storage

Reviewer: ```suggestion
    accelerator enabling flexible and efficient data center networking, security and storage
```

>     efficient and flexible. It offloads from server CPU a set of tasks such as

Reviewer: ```suggestion
    It offloads from server CPU a set of tasks such as
```

    how to process and route packets of data.

>   - **SmartSwitches**. A smart switch is a type of managed switch with a
>     selected number of options for management. 

Reviewer: ```suggestion
  - **SmartSwitches**. A smart switch is a network switch with programmable DPUs/IPUs/EPUs, etc...) SmartNIC hardware integrated into the switch design, with a selected number of options for management. 
```

 that DASH adds a new **SDN control plane** via **gNMI** in the **DASH
 container**. The following figure depicts the DASH software stack.

![dash-software-stack](./images/hld/dash-hld-software-stack.svg)

<figcaption><i>Figure 2 - DASH software stack</i></figcaption><br/>

#### SDN controller

The SDN controller is **primarily responsible for controlling the DASH overlay
services**, while the traditional SONiC application containers are used to
manage the underlay (L3 routing) and hardware platform. Both the DASH container
and the traditional SONiC application containers sit atop the Switch State
services (SWSS) layer, and manipulate the Redis application-layer DBs; these in
turn are translated into SAI dataplane obects via the normal SONiC orchestration
daemons inside SWSS.

The SDN controller controls the overlay built on top of the physical layer
(underlay) of the infrastructure.  From the point of view of the SDN control
plane, when a customer creates an operation from the cloud portal (for example a
VNET creation), the controller allocates the resources, placement management,
capacity, etc. via the  **NorthBound interface APIs**.

##### SDN and DPU High-Availability (HA)

> For **High Availability** (HA), the SDN controller selects a pair of cards and

Reviewer: ```suggestion
High Availability (HA) is important for hardware failover incidents. The SDN controller selects a pair of cards and
```

configures them identically.  The only requirement on the card from the HA
perspective is for the cards to setup a channel between themselves for flow
synchronization.  The synchronization mechanism is left for technology suppliers
to define and implement. For more information, see [High Availability and
Scale](../../high-avail/design/high-availability-and-scale.md) document.

### DASH container

The SDN controller communicates with a DASH device through a
**[gNMI](https://github.com/Azure/DASH/wiki/Glossary#gnmi) endpoint** served by
a new DASH SDN agent **running inside a new SONiC DASH container**.  

In summary:

- The DASH container translates SDN configuration modeled in gNMI into **SONiC
  DB** objects. The gNMI schema is closely related to the DASH DB schema so in

>   effect, the gNMI server is a a thin RPC shim layer to the DB.

Reviewer: ```suggestion
  effect, the gNMI server is a thin RPC shim layer to the DB.
```

- The **SONiC orchagent** inside the Switch State Service (SWSS) Container will
  be enhanced to transform and translate these objects into **SAI_DB objects**,
  including the new **DASH-specific SAI objects**.  
- An **enhanced syncd** will then configure the dataplane using the **technology
  supplier-specific SAI library**.

A **gNMI schema** will manage the following DASH components:

- Elastic Network Interface (ENI)
- Access Control Lists (ACLs)
- Routing and mappings
- Encapsulations
- Other  

> [!NOTE] See also work in progress PR [doc-ref-config-example
> #93](https://github.com/Azure/DASH/pull/93).

#### Multiple DPUs device

In the case of a multiple DPUs device the following applies:

- Each DPU provides a gNMI endpoint for SDN controller through a unique IP
  address.
- An appliance or smart switch containing multiple DPUs will contain multiple
  gNMI endpoints for SDN controller, and the controller treats each DPU as a

>   separate entity.

Reviewer: > [!NOTE]  Note:  Check w/SDN and SONiC that the controller communicates to the DPU, or is it the SONiC switch NOS communicating with the DPU?

Author: Added note to the content.

 wiki and reproduced in diagram form under the [Detailed
 Architectures](#detailed-architectures) section of this document.

### Switch State Service (SWSS)

The SWSS container comprises many daemons which operate on conceptual SONIC
config objects across several databases. DASH affects the following daemons, as
follows:

- `orchagent`, which translates `XX_DB` objects (application and state DBs -
  **TODO** - identify) into `ASIC_DB` objects, must be enhanced to manage new
  DASH overlay objects, such as ACL1,2,3 rules, ENI mappings, etc. The
  `orchagent` has to manage both representations of SONiC objects (`XX_DB` and
  `ASIC_DB`) and translates between them bidirectionally as required.
- `syncd`, which translates `ASIC_DB` conceptual objects into technology
  supplier SAI library API calls, must likewise be enhanced to handle new DASH
  SAI objects.

### Switch Abstraction Interface (SAI) DASH extension

The Switch Abstraction Interface (SAI) is a common API that is supported by many
switch ASIC technology suppliers. SONiC uses SAI to program the ASIC. This
enables SONiC to work across multiple ASIC platforms naturally. DASH uses a

> combination of traditional SAI headers and new DASH pipeline-specific extension headers.

Reviewer: ```suggestion
combination of traditional SAI headers and new DASH pipeline-specific *extension* headers.
```

Technology suppliers must implement this interface for their DASH devices. This
is the primary integration point of DASH devices and the SONiC stack. It will be
rigorously tested for performance and conformance. See [DASH Testing
documentation](https://github.com/Azure/DASH/tree/main/test).

SAI "schema" are represented as fixed c-language header files and derived
metadata header files. The underlay and overlay schema have different origins:

- Traditional SAI headers are defined in the [OCP SAI project
  repo](https://github.com/opencomputeproject/SAI/tree/master/inc).These are

>   hand-generated and maintained. DASH uses a subset of these to manage underlay

Reviewer: ```suggestion
  hand-generated and maintained. DASH uses a subset of these to manage "underlay"
```

  functions, e.g. device management, Layer 3 routing and so forth.
- DASH SAI "overlay" objects are derived from a [P4 Behavioral
  Model](https://github.com/Azure/DASH/tree/main/sirius-pipeline). A script
  reads the P4 model and generates SAI header files.

DASH uses an **enhanced syncd** to configure the dataplane using the technology
supplier-specific SAI library.

### ASIC Drivers

The term "ASIC Drivers" is borrowed from traditional SONiC and SAI, where a
datacenter switch was implemented almost entirely inside an ASIC (or multiple
ASICs). These devices are programmed using a technology supplier Software
Development Kit (SDK) which includes device drivers, kernel modules, etc.

A contemporary DASH "SmartNIC" may consist of many complex hardware components
including multi-core System On A Chip (SoC) ASICs, and the associated software.
For simplicity, the software for such systems which interfaces to the SAI layer
is collectively called the "ASIC driver." More importantly, the technology
supplier SAI library will hide all details and present a uniform interface.

> Dedicated group was looking at PINS (which competes w/DASH). 

Reviewer: ```suggestion
Note:  a dedicated group has been investigating the use of PINS.  

Why DASH and Why Not PINS?  PINS is based upon a specific language (P4), and DPUs on the market would need to build a cross compiler. PINS places the onus onto the Cloud Providers to code and string together the PINS API and their own implementation to create the actions they want to achieve; however it may not be optimal for performance reasons in our project.
```

> Why DASH & Why Not PINS? Reason for DASH is that PINS is based upon a specific language (P4), 
> and DPUs on the market would need to build a cross compiler. 
> SAI is 1 layer above where you can have PINS, or Vendor APIs under the covers. 

Reviewer: ```suggestion

```

> DASH is different b/c all other vendors are describing general purpose APIs to perform general purpose actions (ACL, Forwarding table, Meters). THIS IS THE PROBLEM - it's too generic. They do not know how the APIs will be used, and they don't have knowledge of what the cloud is actually 'doing'. DASH is publicizing Behavioral Models to optimize exactly what we are doing, to optimize against 'what' we do. This is the first time 'what we do' has ever been exposed.

Reviewer: ```suggestion

```
